test/unit/transform/classes_test.py

```diff
@@ -0,0 +1,472 @@
+    [
+        pytest.param(
+            dict(invalid_values=[0, pd.NA], required_valid_cols=["a", "b"]),
+            does_not_raise(),
+            id="good_required_valid_cols",
+        ),
+        pytest.param(
+            dict(invalid_values=[0, pd.NA], allowed_invalid_cols=["a", "b"]),
+            does_not_raise(),
+            id="good_allowed_invalid_cols",
+        ),
+        pytest.param(
+            dict(
+                invalid_values=[0, pd.NA],
+                allowed_invalid_cols=["a", "b"],
+                required_valid_cols=["a", "b"],
+            ),
+            pytest.raises(ValidationError),
+            id="too_many_filters",
+        ),
+        pytest.param(
+            dict(
+                invalid_values=[0, pd.NA],
+                like="dude",
+                regex="wtf",
+            ),
+            pytest.raises(ValidationError),
+            id="too_many_filters",
+        ),
+        pytest.param(
+            dict(
+                invalid_values=[0, pd.NA],
+                required_valid_cols=["a", "b"],
+                like="omg",
+            ),
+            pytest.raises(ValidationError),
+            id="too_many_filters",
+        ),
+        pytest.param(
+            dict(invalid_values=[0, pd.NA]),
+            pytest.raises(ValidationError),
+            id="no_filters",
+        ),
+        pytest.param(
+            dict(invalid_values=[], allowed_invalid_cols=["a", "b"]),
+            pytest.raises(ValidationError),
+            id="no_invalid_values_specified",
+        ),
+    ],
+)
+def test_invalid_row_validation(invalid_rows, expectation):
+    """Make sure we catch invalid arguments to the InvalidRows model."""
+    with expectation:
+        _ = InvalidRows(**invalid_rows)
+
+
+#####################################################################################
+# Series transformation unit tests
+# These transform functions take and return single columns.
+#####################################################################################
+@pytest.mark.parametrize(
+    "series,expected,params", [(STRING_DATA.raw, STRING_DATA.norm, FERC1_STRING_NORM)]
+)
+def test_normalize_strings(series: pd.Series, expected: pd.Series, params) -> None:
+    """Test our string normalization function in isolation."""
+    normalized = normalize_strings(series, StringNormalization(**params))
+    assert_series_equal(normalized, expected, check_names=False)
+
+
+@pytest.mark.parametrize(
+    "series,expected,params", [(STRING_DATA.norm, STRING_DATA.cat, ANIMAL_CATS)]
+)
+def test_categorize_strings(series: pd.Series, expected: pd.Series, params) -> None:
+    """Test string categorization function in isolation."""
+    categorized = categorize_strings(series, params=StringCategories(**params))
+    assert_series_equal(categorized, expected, check_names=False)
+
+
+@pytest.mark.parametrize(
+    "series,expected,params",
+    [
+        pytest.param(
+            NUMERICAL_DATA.year,
```

Comment:
Might be good to add some test values that are exactly on the border of the valid range. Also maybe throw in some negative values just for completeness

Reply:
Okay, added a negative year and capacity, and borderline values as well.